accepted/0017-image-slideshow-redesign.md

```diff
@@ -0,0 +1,100 @@
+# Summary
+
+Re-design the image slideshow to remove its current limitations.
+
+
+# Motivation
+
+In its current implementation, the image slideshow can only play a low number of images.
+The user is not made aware of that fact, they just notice that only the first images are shown in the
+slideshow.
+
+It should be able to show all of the selected images.
+
+A minor issue that may be left out of this RFC is the way random playback is behaving. Currently it picks the 
+next image completely random. This may result in some images getting shown multiple times while others may get 
+shown seldom or not at all. The current implementation also suffers from modulo bias which means the depending 
+on the number of images, the first few images may be "favored" by the algorithm and shown more often. The 
+author's expectation for random playback is to play back the whole list once, but in random order (shuffled),
+thus each image is shown exactly once before the first repeat is seen.
+
+
+# Detailed design
+
+## Main design 
+
+* The slideshow gets a list of files the user has configured to play.
+* At the start, the list is copied to a new list, the playback queue.
+  * The reason for this is to support an improvement in the way random playback is handled (this is discussed below).
+  * It also allows to remove invalid/missing images from the queue but keeps the user's intention (play this 
+    file), so at the next opportunity the slideshow may try to handle this file again.
+* There is a cache/preload buffer that stores only the decoded images that are needed in the near-future or that
+  have been used recently (to handle "go back an image").
```

Comment:
Ah, I see why you'd probably want a non-random method then.  I suppose having the option to choose between random and shuffle is acceptable, but I would like it an option if possible.  I'm not sure how much of a pain that is though.

Reply:
If that's the case, then perhaps there should be both a forward and past cache -- where the "cursor" is in the dead center of the cache if that makes sense.  That would be able to accommodate "go back one image" when using random, at least for a specific number of images.

Reply:
The cache holds up to _n_ previous images, the current image, and up to _n_ next images, if possible. So it already does what you describe. My current design has a queue (which may be shuffled) from which the next image is fetched but I can change the design to implement your proposed way of dealing with randomization. The queue wouldn't be needed any more, then.